### ISSUE TITLE
NoSuchElementException from DeferredContentProviderIterator due to data race in HttpContent.advance()

### ISSUE THREAD

**Issue by shamsimam** [eclipse/jetty.project#3753]
## Possible explanation:

`HttpContent`:
https://github.com/eclipse/jetty.project/blob/e1bc35120a6617ee3df052294e433f3a25ce7097/jetty-client/src/main/java/org/eclipse/jetty/client/HttpContent.java#L153-L159

- In HttpContent, at line 155 [`hashNext()`](https://github.com/eclipse/jetty.project/blob/jetty-9.4.18.v20190429/jetty-client/src/main/java/org/eclipse/jetty/client/util/DeferredContentProvider.java#L250) can return true when the `chunks` in `DeferredContentProviderIterator` is empty
- now in another thread, the DeferredContentProvider is closed, causing `chunks` to [contain `CLOSE`](https://github.com/eclipse/jetty.project/blob/jetty-9.4.18.v20190429/jetty-client/src/main/java/org/eclipse/jetty/client/util/DeferredContentProvider.java#L214).
- In HttpContent, at line 156 [`next()`] will now [throw a `NoSuchElementException`](https://github.com/eclipse/jetty.project/blob/jetty-9.4.18.v20190429/jetty-client/src/main/java/org/eclipse/jetty/client/util/DeferredContentProvider.java#L265)

### Potential fix

Make `DeferredContentProviderIterator` return an ~empty~ `null` buffer in this scenario on the call to `next()` if a previous call to `hasNext()` has promised data by returning `true`.

## Stack trace:

Jetty version: 9.4.18.v20190429
Detailed stack trace (from a Clojure project):
```
[{:type java.util.NoSuchElementException
   :message nil
   :at [org.eclipse.jetty.client.util.DeferredContentProvider$DeferredContentProviderIterator next DeferredContentProvider.java 265]}]
 :trace
 [[org.eclipse.jetty.client.util.DeferredContentProvider$DeferredContentProviderIterator next DeferredContentProvider.java 265]
  [org.eclipse.jetty.client.util.DeferredContentProvider$DeferredContentProviderIterator next DeferredContentProvider.java 241]
  [qbits.jet.client.http$wrap_iterator$reify__18786 next http.clj 311]
  [org.eclipse.jetty.client.HttpContent advance HttpContent.java 156]
  [org.eclipse.jetty.client.HttpContent advance HttpContent.java 149]
  [org.eclipse.jetty.client.HttpSender$ContentCallback process HttpSender.java 849]
  [org.eclipse.jetty.util.IteratingCallback processing IteratingCallback.java 241]
  [org.eclipse.jetty.util.IteratingCallback succeeded IteratingCallback.java 365]
  [org.eclipse.jetty.client.HttpSender$ContentCallback succeeded HttpSender.java 907]
  [org.eclipse.jetty.http2.HTTP2Stream succeeded HTTP2Stream.java 530]
  [org.eclipse.jetty.util.Callback$Nested succeeded Callback.java 261]
  [org.eclipse.jetty.http2.HTTP2Session$DataEntry succeeded HTTP2Session.java 1461]
  [java.lang.Iterable forEach Iterable.java 75]
  [org.eclipse.jetty.http2.HTTP2Flusher finish HTTP2Flusher.java 276]
  [org.eclipse.jetty.http2.HTTP2Flusher succeeded HTTP2Flusher.java 268]
  [org.eclipse.jetty.io.WriteFlusher write WriteFlusher.java 293]
  [org.eclipse.jetty.io.AbstractEndPoint write AbstractEndPoint.java 380]
  [org.eclipse.jetty.http2.HTTP2Flusher process HTTP2Flusher.java 247]
  [org.eclipse.jetty.util.IteratingCallback processing IteratingCallback.java 241]
  [org.eclipse.jetty.util.IteratingCallback iterate IteratingCallback.java 224]
  [org.eclipse.jetty.http2.HTTP2Session newStream HTTP2Session.java 569]
  [org.eclipse.jetty.http2.client.http.HttpSenderOverHTTP2 sendHeaders HttpSenderOverHTTP2.java 95]
  [org.eclipse.jetty.client.HttpSender send HttpSender.java 214]
  [org.eclipse.jetty.http2.client.http.HttpChannelOverHTTP2 send HttpChannelOverHTTP2.java 96]
  [org.eclipse.jetty.client.HttpChannel send HttpChannel.java 128]
  [org.eclipse.jetty.client.HttpConnection send HttpConnection.java 201]
  [org.eclipse.jetty.http2.client.http.HttpConnectionOverHTTP2 send HttpConnectionOverHTTP2.java 77]
  [org.eclipse.jetty.http2.client.http.HttpDestinationOverHTTP2 send HttpDestinationOverHTTP2.java 38]
  [org.eclipse.jetty.client.HttpDestination process HttpDestination.java 346]
  [org.eclipse.jetty.client.HttpDestination process HttpDestination.java 304]
  [org.eclipse.jetty.client.HttpDestination send HttpDestination.java 294]
  [org.eclipse.jetty.client.HttpDestination send HttpDestination.java 269]
  [org.eclipse.jetty.client.HttpDestination send HttpDestination.java 246]
  [org.eclipse.jetty.client.HttpClient send HttpClient.java 576]
  [org.eclipse.jetty.client.HttpRequest send HttpRequest.java 726]
  [org.eclipse.jetty.client.HttpRequest send HttpRequest.java 718]
  ...]
```

**Comment by shamsimam** [comment]
I can submit a PR with the suggested fix if it seems acceptable.

**Comment by gregw** [comment]
@shamsimam thanks for this report, but I don't think your proposed solution is correct, because it could result in a loop of something calling `hasNext()` and `next()` getting null each time.
I think the proper fix is to make the `hasNext()` method call `poll()` and thus we will block there for content rather than in `hasNext()`.  The `next()` method should then call hasNext().     Any element taken from the poll will need to be remembered in a field so that a subsequent call to `hasNext` will not `poll` another element until the previous has been consumed.

@sbordet do you agree?

**Comment by shamsimam** [comment]
@gregw The use case for `DeferredContentProvider` is to support asynchronous data. We do not want to block on the call to `hashNext()`.

**Comment by shamsimam** [comment]
@gregw Regarding the loop of `hasNext()` and `next()` and getting null each time:

As per the [javadoc of `HttpContent`](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-client/src/main/java/org/eclipse/jetty/client/HttpContent.java#L61-L64), there should be a call to `hasNext()` only after data has become available.

**Comment by gregw** [comment]
@shamsimam , the implementation of `next()` is already calling poll, so there is still the chance for blocking. `hasNext()` should be called before `next()` so moving the poll there should be OK

**Comment by shamsimam** [comment]
My bad. Calling `poll` is non-blocking, returning `null` when no data is available. However, if data is inserted after that last call to `hasNext()` when no data was available, we should return the new element instead of returning `null` from the remembered field.

**Comment by gregw** [comment]
ah, if null is non blocking, then my idea doesn't work.    Let me chat with @sbordet in the new week to see what should be done.

**Comment by sbordet** [comment]
@shamsimam your stack trace is really strange because it shows this:

```
  [org.eclipse.jetty.client.util.DeferredContentProvider$DeferredContentProviderIterator next DeferredContentProvider.java 241]
  [qbits.jet.client.http$wrap_iterator$reify__18786 next http.clj 311]
  [org.eclipse.jetty.client.HttpContent advance HttpContent.java 156]
  [org.eclipse.jetty.client.HttpContent advance HttpContent.java 149]
```

In particular, `HttpContent:149` is the `else` branch for the check `iterator instanceof Synchronizable`.
It happens to be that `DeferredContentProviderIterator` _does_ implement `Synchronizable`, so I don't understand why the `else` branch is taken.
Is clojure doing some wrapping to the iterator so that the `instanceof` check fails?

**Comment by gregw** [comment]
@sbordet regardless of the calling context and other strangeness, I think the iterator here is just wrong and should be something like:
```java
        private Chunk current;

        @Override
        public boolean hasNext()
        {
            synchronized (lock)
            {
                if (current == null)
                    current = chunks.poll();
                return current != null && current != CLOSE;
            }
        }

        @Override
        public ByteBuffer next()
        {
            synchronized (lock)
            {
                if (!hasNext())
                    throw new NoSuchElementException();
                Chunk chunk = current;
                current = null;
                return chunk.buffer;
            }
        }
```

**Comment by sbordet** [comment]
@gregw this is a deferred content iterator, so when it has no content (i.e. `chunks.poll()` returns `null`) the `hasNext()` must return `true` because it has content, just that it is not yet present.
And yes I regret choosing `Iterator` as the abstraction for content - there are so many cases that an iterator does not cover and I should have really used a better abstraction.
Perhaps revisit this for Jetty 10, but for now I need `hasNext()` to return false only and only if it sees the `CLOSE` pill - I filed #3756.

**Comment by gregw** [comment]
@sbordet Hmmm  not convinced.    If this iterator should never be called when chunks is empty, then a poll() returning null prior to a CLOSE should result in a ISE.   Turning that into true, just exposes the race this issue is about.

**Comment by sbordet** [comment]
This iterator can be called and return `null`. That is how the implementation knows that it's async content and relinquishes the thread back to the pool.

```
hasNext = true, next = chunk => we have content to process now
hasNext = true, next = null => we don't have content to process now, but it will arrive
hasNext = false, next = CLOSE => the content is finished (so e.g. put `endStream=true` in HTTP/2)
```

**Comment by shamsimam** [comment]
@sbordet 

> Is clojure doing some wrapping to the iterator so that the` instanceof` check fails?

This is introduced our code, the [wrapper](https://github.com/twosigma/jet/blob/waiter-patch/src/clj/qbits/jet/client/http.clj#L292)  does not implement `Synchronizable`. I will fix that at our end for the iterator to implement `Synchronizable`. 

Is it required that iterators of `AsyncContentProvider` implement the `Synchronizable` interface?

**Comment by sbordet** [comment]
@shamsimam wow that's accessing a lot of Jetty internals that were not supposed to be accessed.
Can I ask why you need to wrap the iterator? From what I grok of Clojure, seems to me it is doing exactly the same that the Jetty implementation is doing (also in other `fn`s).

To answer your question, it is not strictly requires that an `AsyncContentProvider` implementation also implement `Synchronizable`, but in your case is it necessary.
But it should not be necessary that you do all that wrapping, and if you don't even have the problem.

**Comment by shamsimam** [comment]
@sbordet 

> Can I ask why you need to wrap the iterator?
>  it should not be necessary that you do all that wrapping

We would love to avoid the wrapping.
The wrapper is to get a callback on when the [last content has been produced](https://github.com/twosigma/jet/blob/waiter-patch/src/clj/qbits/jet/client/http.clj#L305).
We use it to determine when it is safe to try and [read the trailers from the request](https://github.com/twosigma/jet/blob/waiter-patch/src/clj/qbits/jet/client/http.clj#L428).

**Comment by sbordet** [comment]
@shamsimam I don't understand. Application code does not read the trailers, it sets it.
It's the implementation that reads the request trailers at the right time and send them appropriately.
You can implement all your late logic inside the `Supplier<HttpFields>` (e.g. producing a hash of the content).

**Comment by shamsimam** [comment]
Setting a `Supplier` means that, for HTTP/2, an empty trailers frame is sent if the logic determines that there are no trailers to send. This is problematic because the gRPC servers fail whenever they see a trailers frame (even if it is empty) on incoming requests. We use the wrapper to avoid setting the `Supplier` and hence avoid sending the empty trailers frame.

**Comment by sbordet** [comment]
That is a gRPC server bug, it cannot reject an empty frame as those are completely allowed by the specification.

And the Jetty client tries as hard as possible to _not_ send an empty frame, so again I don't know in what exact case you see this empty frame being sent.
We may have a bug there, but from a look at the code seems to me we're not sending an empty frame if there are no trailers.

**Comment by shamsimam** [comment]
> That is a gRPC server bug, it cannot reject an empty frame as those are completely allowed by the specification.

I agree it is a gRPC server bug. [gRPC-http/2 specification](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md) says it does not expect trailers in requests. The gRPC server may have read this and chosen to not accept trailer frames in incoming requests.

> We may have a bug there, but from a look at the code seems to me we're not sending an empty frame if there are no trailers.

The need to have to send a trailers frame probably comes from one of these where `lastContent` evaluates to `false` and a trailers needs to be sent to end the stream:
- https://github.com/eclipse/jetty.project/blob/master/jetty-http2/http2-http-client-transport/src/main/java/org/eclipse/jetty/http2/client/http/HttpSenderOverHTTP2.java#L73
- https://github.com/eclipse/jetty.project/blob/master/jetty-http2/http2-http-client-transport/src/main/java/org/eclipse/jetty/http2/client/http/HttpSenderOverHTTP2.java#L123

**Comment by sbordet** [comment]
> The need to have to send a trailers frame probably comes from one of these where lastContent evaluates to false and a trailers needs to be sent to end the stream:

I don't understand. The case you mentions will send an empty data frame only if the chunk provided by the application is a zero-length `ByteBuffer`.

So if gRPC does not allow request trailers, why do you send them?

**Comment by joakime** [comment]
Adding commit specific links of lines referenced by @shamsimam to maintain history for those looking at this discussion in the future.

https://github.com/eclipse/jetty.project/blob/0e282eece2c32a16b24ca22b994b5327212a7425/jetty-http2/http2-http-client-transport/src/main/java/org/eclipse/jetty/http2/client/http/HttpSenderOverHTTP2.java#L73

https://github.com/eclipse/jetty.project/blob/0e282eece2c32a16b24ca22b994b5327212a7425/jetty-http2/http2-http-client-transport/src/main/java/org/eclipse/jetty/http2/client/http/HttpSenderOverHTTP2.java#L123

**Comment by shamsimam** [comment]
>>> You can implement all your late logic inside the Supplier<HttpFields>

>> The need to have to send a trailers frame probably comes from...

> So if gRPC does not allow request trailers, why do you send them?

Our application is a generic http proxy that needs to work for all http/2 requests. I was referring to Jetty behavior where it sends an empty _trailers_ frame if the `trailers` field is non-`null`. As a result, we cannot use a `Supplier` and lazily decide whether or not to send trailers for gRPC requests. Using the wrapping code that we have, we lazily set the `trailers` field and can avoid having Jetty send an empty _trailers_ frame when that field is `null`, instead ending the stream on the _data_ frame as the last content.

**Comment by sbordet** [comment]
So what are you asking is that we also treat empty trailers _fields_ like a `null` trailer _supplier_ and avoid sending an empty frame if the supplier is non-`null` but the trailer fields are empty?

**Comment by shamsimam** [comment]
> So what are you asking is that we also treat empty trailers fields like a null trailer supplier and avoid sending an empty frame if the supplier is non-null but the trailer fields are empty?

If possible, yes. That would allow us to avoid the wrapping logic that we have.  I'm not sure if it needs its own issue.

This issue is about `NoSuchElementException ` from `DeferredContentProviderIterator` (admittedly caused by the wrapping). As you suggested, a fix is for the wrapper to also implement `Synchronizable`.

**Comment by sbordet** [comment]
@shamsimam yes please close this issue and open a new one.

**Comment by shamsimam** [comment]
Created #3758. I will close this issue after I have verified that the `Synchronizable` approach works.

**Comment by shamsimam** [comment]
Implementing `Synchronized` (as suggested) fixed the issue.
